uc-crux-llvm/src/UCCrux/LLVM/View/Shape.hs

```diff
@@ -0,0 +1,197 @@
+{-
+Module           : UCCrux.LLVM.View.Shape
+Description      : See "UCCrux.LLVM.View".
+Copyright        : (c) Galois, Inc 2022
+License          : BSD3
+Maintainer       : Langston Barrett <langston@galois.com>
+Stability        : provisional
+-}
+
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StrictData #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module UCCrux.LLVM.View.Shape
+  ( ViewShapeError,
+    ppViewShapeError,
+    PtrShapeView(..),
+    ptrShapeView,
+    viewPtrShape,
+    ShapeView(..),
+    shapeView,
+    viewShape,
+  )
+where
+
+import           Control.Monad (when)
+import qualified Data.Aeson as Aeson
+import qualified Data.Aeson.TH as Aeson.TH
+import           Data.Foldable (toList)
+import           Data.List.NonEmpty (NonEmpty, nonEmpty)
+import           Data.Sequence (Seq)
+import           Data.Vector (Vector)
+import qualified Data.Vector as Vec
+import           GHC.Generics (Generic)
+import           Numeric.Natural (Natural)
+
+import           Prettyprinter (Doc)
+
+import qualified Data.Parameterized.Context as Ctx
+import qualified Data.Parameterized.NatRepr as NatRepr
+import           Data.Parameterized.TraversableFC (toListFC)
+import           Data.Parameterized.TraversableFC.WithIndex (itraverseFC)
+import qualified Data.Parameterized.Vector as PVec
+
+import           UCCrux.LLVM.Errors.Panic (panic)
+import           UCCrux.LLVM.FullType.Type (FullTypeRepr(..), ModuleTypes, asFullType)
+import           UCCrux.LLVM.Shape (PtrShape(..), Shape(..))
+
+data ViewShapeError e
+  = TagError e
+  | TypeMismatch
+  | StructLengthMismatch
+  | VectorLengthMismatch !Natural !Int
+  deriving (Eq, Ord, Show)
+
+ppViewShapeError :: (e -> Doc ann) -> ViewShapeError e -> Doc ann
+ppViewShapeError err =
+  \case
+    TagError e -> err e
+    TypeMismatch -> "Type mismatch"
+    StructLengthMismatch -> "Struct length mismatch"
+    VectorLengthMismatch{} -> "Vector length mismatch"
+
+-- Helper, not exported
+liftErr :: Either e a -> Either (ViewShapeError e) a
+liftErr =
+  \case
+    Left e -> Left (TagError e)
+    Right v -> Right v
+
+data PtrShapeView vtag
+  = VShapeUnallocated
+  | VShapeAllocated Int
+  | VShapeInitialized (Seq (ShapeView vtag))
+  deriving (Eq, Generic, Ord, Show)
+
+ptrShapeView ::
+  (forall t. tag t -> vtag) ->
+  PtrShape m tag ft ->
+  PtrShapeView vtag
+ptrShapeView tag =
+  \case
+    ShapeUnallocated -> VShapeUnallocated
+    ShapeAllocated n -> VShapeAllocated n
+    ShapeInitialized s -> VShapeInitialized (fmap (shapeView tag) s)
+
+viewPtrShape ::
+  ModuleTypes m ->
+  (forall t. FullTypeRepr m t -> vtag -> Either e (tag t)) ->
```

Comment:
Most of the other arguments are pretty straightforward. I think this one is sufficiently complex that it would be worth documenting roughly what this means and what kinds of values a caller might want to pass

Reply:
Added a comment!